File: src/selfevals/api/routes/metrics.py
"""Aggregate metrics: pass-rate, failure-modes, tools, cost, tokens, latency."""

from __future__ import annotations

from datetime import datetime
from typing import Annotated

from fastapi import Depends, FastAPI, Query

from selfevals.api.auth import UserHeader
from selfevals.api.deps import AppDeps
from selfevals.api.metrics import (
    cost_metrics,
    failure_mode_metrics,
    latency_metrics,
    pass_rate_metrics,
    token_metrics,
    tool_metrics,
)
from selfevals.api.schemas import (
    CostMetricsResponse,
    FailureModeMetricsResponse,
    LatencyMetricsResponse,
    PassRateMetricsResponse,
    TokenMetricsResponse,
    ToolMetricsResponse,
)
from selfevals.storage.interface import StorageInterface


def register(app: FastAPI, deps: AppDeps) -> None:
    _register_quality_metrics(app, deps)
    _register_cost_metrics(app, deps)


def _register_quality_metrics(app: FastAPI, deps: AppDeps) -> None:
    @app.get(
        "/api/workspaces/{workspace_id}/metrics/pass-rate",
        response_model=PassRateMetricsResponse,
        tags=["metrics"],
    )
    def metrics_pass_rate(
        workspace_id: str,
        storage: StorageInterface = Depends(deps.storage),
        start: Annotated[datetime | None, Query(alias="from")] = None,
        end: Annotated[datetime | None, Query(alias="to")] = None,
        experiment_id: str | None = None,
        grader: str | None = None,
        _user: UserHeader = None,
    ) -> PassRateMetricsResponse:
        try:
            return pass_rate_metrics(
                storage,
                workspace_id=workspace_id,
                start=start,
                end=end,
                experiment_id=experiment_id,
                grader=grader,
            )
        finally:
            storage.close()

    @app.get(
        "/api/workspaces/{workspace_id}/metrics/failure-modes",
        response_model=FailureModeMetricsResponse,
        tags=["metrics"],
    )
    def metrics_failure_modes(
        workspace_id: str,
        storage: StorageInterface = Depends(deps.storage),
        start: Annotated[datetime | None, Query(alias="from")] = None,
        end: Annotated[datetime | None, Query(alias="to")] = None,
        experiment_id: str | None = None,
        grader: str | None = None,
        _user: UserHeader = None,
    ) -> FailureModeMetricsResponse:
        try:
            return failure_mode_metrics(
                storage,
                workspace_id=workspace_id,
                start=start,
                end=end,
                experiment_id=experiment_id,
                grader=grader,
            )
        finally:
            storage.close()

    @app.get(
        "/api/workspaces/{workspace_id}/metrics/tools",
        response_model=ToolMetricsResponse,
        tags=["metrics"],
    )
    def metrics_tools(
        workspace_id: str,
        storage: StorageInterface = Depends(deps.storage),
        start: Annotated[datetime | None, Query(alias="from")] = None,
        end: Annotated[datetime | None, Query(alias="to")] = None,
        experiment_id: str | None = None,
        tool_name: str | None = None,
        _user: UserHeader = None,
    ) -> ToolMetricsResponse:
        try:
            return tool_metrics(
                storage,
                workspace_id=workspace_id,
                start=start,
                end=end,
                experiment_id=experiment_id,
                tool_name=tool_name,
            )
        finally:
            storage.close()


def _register_cost_metrics(app: FastAPI, deps: AppDeps) -> None:
    @app.get(
        "/api/workspaces/{workspace_id}/metrics/cost",
        response_model=CostMetricsResponse,
        tags=["metrics"],
    )
    def metrics_cost(
        workspace_id: str,
        storage: StorageInterface = Depends(deps.storage),
        start: Annotated[datetime | None, Query(alias="from")] = None,
        end: Annotated[datetime | None, Query(alias="to")] = None,
        experiment_id: str | None = None,
        model: str | None = None,
        _user: UserHeader = None,
    ) -> CostMetricsResponse:
        try:
            return cost_metrics(
                storage,
                workspace_id=workspace_id,
                start=start,
                end=end,
                experiment_id=experiment_id,
                model=model,
            )
        finally:
            storage.close()

    @app.get(
        "/api/workspaces/{workspace_id}/metrics/tokens",
        response_model=TokenMetricsResponse,
        tags=["metrics"],
    )
    def metrics_tokens(
        workspace_id: str,
        storage: StorageInterface = Depends(deps.storage),
        start: Annotated[datetime | None, Query(alias="from")] = None,
        end: Annotated[datetime | None, Query(alias="to")] = None,
        experiment_id: str | None = None,
        model: str | None = None,
        _user: UserHeader = None,
    ) -> TokenMetricsResponse:
        try:
            return token_metrics(
                storage,
                workspace_id=workspace_id,
                start=start,
                end=end,
                experiment_id=experiment_id,
                model=model,
            )
        finally:
            storage.close()

    @app.get(
        "/api/workspaces/{workspace_id}/metrics/latency",
        response_model=LatencyMetricsResponse,
        tags=["metrics"],
    )
    def metrics_latency(
        workspace_id: str,
        storage: StorageInterface = Depends(deps.storage),
        start: Annotated[datetime | None, Query(alias="from")] = None,
        end: Annotated[datetime | None, Query(alias="to")] = None,
        experiment_id: str | None = None,
        _user: UserHeader = None,
    ) -> LatencyMetricsResponse:
        try:
            return latency_metrics(
                storage,
                workspace_id=workspace_id,
                start=start,
                end=end,
                experiment_id=experiment_id,
            )
        finally:
            storage.close()

File: web/src/lib/api/resources/metrics.ts
import { qs, request } from '$lib/api/http';
import type {
  AnalysisBundle,
  AnalysisIngestSummary,
  AnalysisResult,
  CostMetrics,
  FailureClusters,
  FailureMode,
  FailureModeMetrics,
  LatencyMetrics,
  PassRateMetrics,
  TokenMetrics,
  ToolMetrics
} from '$lib/api/types';

export const metricsApi = {
  // Each takes an optional `from`/`to` (ISO 8601) window + per-metric filter.

  metricsPassRate: (
    workspaceId: string,
    opts: { from?: string; to?: string; experiment_id?: string; grader?: string } = {},
    fetch?: typeof globalThis.fetch
  ) =>
    request<PassRateMetrics>(`/api/workspaces/${workspaceId}/metrics/pass-rate${qs(opts)}`, {
      fetch
    }),

  metricsFailureModes: (
    workspaceId: string,
    opts: { from?: string; to?: string; experiment_id?: string; grader?: string } = {},
    fetch?: typeof globalThis.fetch
  ) =>
    request<FailureModeMetrics>(`/api/workspaces/${workspaceId}/metrics/failure-modes${qs(opts)}`, {
      fetch
    }),

  metricsTools: (
    workspaceId: string,
    opts: { from?: string; to?: string; experiment_id?: string; tool_name?: string } = {},
    fetch?: typeof globalThis.fetch
  ) => request<ToolMetrics>(`/api/workspaces/${workspaceId}/metrics/tools${qs(opts)}`, { fetch }),

  metricsCost: (
    workspaceId: string,
    opts: { from?: string; to?: string; experiment_id?: string; model?: string } = {},
    fetch?: typeof globalThis.fetch
  ) => request<CostMetrics>(`/api/workspaces/${workspaceId}/metrics/cost${qs(opts)}`, { fetch }),

  metricsTokens: (
    workspaceId: string,
    opts: { from?: string; to?: string; experiment_id?: string; model?: string } = {},
    fetch?: typeof globalThis.fetch
  ) => request<TokenMetrics>(`/api/workspaces/${workspaceId}/metrics/tokens${qs(opts)}`, { fetch }),

  metricsLatency: (
    workspaceId: string,
    opts: { from?: string; to?: string; experiment_id?: string } = {},
    fetch?: typeof globalThis.fetch
  ) =>
    request<LatencyMetrics>(`/api/workspaces/${workspaceId}/metrics/latency${qs(opts)}`, {
      fetch
    }),

  /**
   * Failing traces grouped by failure mode (§J.6). A first-order view (not under
   * `/metrics`) — each cluster carries example `run_id`s that link straight into
   * the trace viewer. v1 clusters by the stable taxonomy slug.
   */
  clusters: (
    workspaceId: string,
    opts: {
      from?: string;
      to?: string;
      experiment_id?: string;
      grader?: string;
      limit?: number;
    } = {},
    fetch?: typeof globalThis.fetch
  ) =>
    request<FailureClusters>(`/api/workspaces/${workspaceId}/clusters${qs(opts)}`, {
      fetch
    }),

  // --- Failure-mode taxonomy (2A) --------------------------------------

  listFailureModes: (
    workspaceId: string,
    opts: { status?: string } = {},
    fetch?: typeof globalThis.fetch
  ) =>
    request<{ items: FailureMode[] }>(`/api/workspaces/${workspaceId}/failure-modes${qs(opts)}`, {
      fetch
    }),

  promoteFailureMode: (workspaceId: string, id: string, fetch?: typeof globalThis.fetch) =>
    request<FailureMode>(`/api/workspaces/${workspaceId}/failure-modes/${id}/promote`, {
      method: 'POST',
      fetch
    }),

  retireFailureMode: (workspaceId: string, id: string, fetch?: typeof globalThis.fetch) =>
    request<FailureMode>(`/api/workspaces/${workspaceId}/failure-modes/${id}/retire`, {
      method: 'POST',
      fetch
    }),

  mergeFailureMode: (
    workspaceId: string,
    id: string,
    intoId: string,
    fetch?: typeof globalThis.fetch
  ) =>
    request<FailureMode>(`/api/workspaces/${workspaceId}/failure-modes/${id}/merge`, {
      method: 'POST',
      json: { into_id: intoId },
      fetch
    }),

  editFailureMode: (
    workspaceId: string,
    id: string,
    patch: { title?: string; definition?: string },
    fetch?: typeof globalThis.fetch
  ) =>
    request<FailureMode>(`/api/workspaces/${workspaceId}/failure-modes/${id}`, {
      method: 'PATCH',
      json: patch,
      fetch
    }),

  // --- Error-analysis bundle / ingest (2C) -----------------------------

  analysisBundle: (
    workspaceId: string,
    experimentId: string,
    opts: { iteration?: number; all?: boolean } = {},
    fetch?: typeof globalThis.fetch
  ) =>
    request<AnalysisBundle>(
      `/api/workspaces/${workspaceId}/experiments/${experimentId}/analysis/bundle${qs({
        iteration: opts.iteration,
        all: opts.all ? 'true' : undefined
      })}`,
      { fetch }
    ),

  analysisIngest: (
    workspaceId: string,
    experimentId: string,
    result: AnalysisResult,
    fetch?: typeof globalThis.fetch
  ) =>
    request<AnalysisIngestSummary>(
      `/api/workspaces/${workspaceId}/experiments/${experimentId}/analysis/ingest`,
      { method: 'POST', json: result, fetch }
    )
};
